Rewrite in pure node to support Windows.

calavera opened the pull request:
It adds two dependencies, but I think it's worth making it work for any OS.

Fixes #1 

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Rewrite in pure node to support Windows.

It adds two dependencies, but I think it's worth making it work for any OS.

Signed-off-by: David Calavera <david.calavera@gmail.com>

jpaquim commented on packages/cli/index.js:
```diff
@@ -3,7 +3,12 @@
 import * as path from 'path';
 import * as url from 'url';
-import { $, ProcessOutput } from 'zx';
+import download from 'download';
```
Mostly a matter of personal taste, but I try to keep node's internal libs listed first, so I suggest something like:
```js
import { exec, spawn } from 'child_process';
import * as fs from 'fs/promises';
import * as path from 'path';
import * as url from 'url';
import { promisify } from 'util';
import download from 'download';
import which from 'which';

const execAsync = promisify(exec);
const spawnAsync = promisify(spawn);
```

jpaquim commented on packages/cli/index.js:
```diff
@@ -61,16 +65,61 @@ export async function run(...args) {
+      .filter(dir => !dir.endsWith(path.join('node_modules', '.bin')))
+      .join(':');
+    // check if there's already an installed zig binary
+    const systemZig = await which('zig');
```
From the `which` docs, I think instead of overwriting the `process.env.PATH`, we can use the extra path argument instead, what do you think? https://github.com/npm/node-which#options

jpaquim commented on packages/cli/index.js:
```diff
@@ -61,16 +65,61 @@ export async function run(...args) {
 
   try {
-    await $`${binaryPath} ${args}`;
+    await spawnAsync(binaryPath, args, { stdio: 'inherit' });
```
Not sure if this will actually work correctly, it doesn't seem like `spawn` takes a callback directly, so I think it's not really being promisified. From the few tests I did locally, it seemed like the Promise was never resolving

jpaquim replied:
Not sure if this will actually work correctly, it doesn't seem like `spawn` takes a callback directly, so I think it's not really being promisified. From the few tests I did locally, it seemed like the Promise was never resolving

jpaquim commented on packages/cli/package.json:
```diff
@@ -32,7 +33,8 @@
     "prepublish": "rm -rf bin && git checkout bin"
   },
   "dependencies": {
```
Can you re-run `pnpm i` to update the `pnpm-lock.yaml` file with the new dependencies?

jpaquim reviewed (commented):
Overall looks pretty good! Thanks for the PR 😄

calavera committed:
Fix Run function

- Make function synchronous (because it doesn't matter in this situation).
- Reorder imports.
- Use path option in `which`.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Update PNPM.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera commented:
@jpaquim I've made those updates, let me know what you think.

jpaquim commented:
LGTM, have you had the chance to test on Windows? I just tested on Linux and it doesn't seem to be working, the tar.xz file seems to be downloaded to the bin folder, but the binary is not extracted, and then the zig.js script always fails. It also didn't seem to detect my system zig correctly...

I can look into the code tomorrow and try to see where it may be failing, but let me know if you happen to have a clue on what may be causing this.

calavera committed:
Fix pathSeparator.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Fix installation on Linux

Add necessary plugins to uncompress tar.xz files.
Fix path variable.

Signed-off-by: David Calavera <david.calavera@gmail.com>

jpaquim commented:
Hey @calavera, not sure if you had the chance to further develop/test this, I can take another look through the PR if you want 🙂

calavera commented:
nope, I ended up not using it.
